Sources/NIOPosix/SelectorWSAPoll.swift

```diff
@@ -0,0 +1,178 @@
+            self.formUnion(.read)
+        }
+        if mapped & WinSDK.POLLWRNORM != 0 {
+            self.formUnion(.write)
+        }
+        if mapped & WinSDK.POLLERR != 0 {
+            self.formUnion(.error)
+        }
+        if mapped & WinSDK.POLLHUP != 0 {
+            self.formUnion(.reset)
+        }
+        if mapped & WinSDK.POLLNVAL != 0 {
+            preconditionFailure("Invalid fd supplied.")
+        }
+    }
+}
+
+extension Selector: _SelectorBackendProtocol {
+
+    func initialiseState0() throws {
+        self.pollFDs.reserveCapacity(16)
+    }
+
+    func deinitAssertions0() {
+        // no global state. nothing to check
+    }
+
+    @inlinable
+    func whenReady0(
+        strategy: SelectorStrategy,
+        onLoopBegin: () -> Void,
+        _ body: (SelectorEvent<R>) throws -> Void
+    ) throws {
+        let time: Int32 =
+            switch strategy {
+            case .now:
+                0
+
+            case .block:
+                -1
+
+            case .blockUntilTimeout(let timeAmount):
+                Int32(clamping: timeAmount.nanoseconds / 1_000_000)
+            }
+
+        let result = self.pollFDs.withUnsafeMutableBufferPointer { ptr in
+            WSAPoll(ptr.baseAddress!, UInt32(ptr.count), time)
+        }
+
+        if result > 0 {
+            // something has happened
+            for i in self.pollFDs.indices {
+                let pollFD = self.pollFDs[i]
+                guard pollFD.revents != 0 else {
+                    continue
+                }
+                // reset the revents
+                self.pollFDs[i].revents = 0
+                let fd = pollFD.fd
+
+                // If the registration is not in the Map anymore we deregistered it during the processing of whenReady(...). In this case just skip it.
+                guard let registration = registrations[Int(fd)] else {
+                    continue
+                }
+
+                var selectorEvent = SelectorEventSet(wsaEvent: pollFD.revents)
+                // in any case we only want what the user is currently registered for & what we got
+                selectorEvent = selectorEvent.intersection(registration.interested)
+
+                guard selectorEvent != ._none else {
+                    continue
+                }
+
+                try body((SelectorEvent(io: selectorEvent, registration: registration)))
+            }
+        } else if result == 0 {
+            // nothing has happened
+        } else if result == WinSDK.SOCKET_ERROR {
+            throw IOError(winsock: WSAGetLastError(), reason: "WSAPoll")
+        }
+    }
+
+    func register0(
+        selectableFD: NIOBSDSocket.Handle,
+        fileDescriptor: NIOBSDSocket.Handle,
+        interested: SelectorEventSet,
+        registrationID: SelectorRegistrationID
+    ) throws {
+        let poll = pollfd(fd: UInt64(fileDescriptor), events: interested.wsaPollEventSet, revents: 0)
+        self.pollFDs.append(poll)
```

Comment:
Totally acceptable starting implementation, but we'll want a different data-structure so that these operations aren't linear time. NIO definitely assumes they're gonna be constant-time.

Reply:
added comment.